app/src/main/java/com/celzero/bravedns/service/ConnectionMonitor.kt

```diff
@@ -0,0 +1,264 @@
+     */
+    fun onPreferenceChanged(){
+        isAllNetworkSupported = persistentState.isAddAllNetwork
+        serviceHandler?.removeMessages(MSG_ADD_ACTIVE_NETWORK, null)
+        serviceHandler?.removeMessages(MSG_ADD_ALL_NETWORKS, null)
+        if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - onPreferenceChanged")
+        if (!isAllNetworkSupported) {
+            val message = constructMessage(serviceHandler?.obtainMessage(), MSG_ADD_ACTIVE_NETWORK)
+            serviceHandler?.sendMessage(message)
+        } else {
+            val message = constructMessage(serviceHandler?.obtainMessage(), MSG_ADD_ALL_NETWORKS)
+            serviceHandler?.sendMessage(message)
+        }
+    }
+
+    fun getNetworkList(): MutableSet<Network> {
+        return currentNetworkList
+    }
+
+    /**
+     * Constructs the message object for Network handler.
+     * Add the active network to the message object
+     * in case of setUnderlying network has only active networks.
+     */
+    private fun constructMessage(what: Int): Message {
+        val message = Message()
+        message.what = what
+        return message
+    }
+
+    private fun constructMessage(message: Message?, what: Int): Message {
+        var messageValue = message
+        if (messageValue == null) {
+            messageValue = Message()
+        }
+        messageValue.what = what
+        return messageValue
+    }
+
+    // Handles the network messages from the callback from the connectivity manager
+    private class NetworkRequestHandler(looper: Looper?, val connectionMonitor: ConnectionMonitor) : Handler(looper!!) {
+        override fun handleMessage(msg: Message) {
+            processMessage(msg)
+        }
+
+        fun processMessage(msg: Message) {
+            if (msg.what == MSG_ADD_ACTIVE_NETWORK) {
+                processActiveNetwork()
+            } else if (msg.what == MSG_ADD_ALL_NETWORKS) {
+                processAllNetwork()
+            }
+        }
+
+        /**
+         * Handles the active network.
+         * Get active network from connectivity-manager
+         * Set the vpn underlying network when the current active network is different from
+         * already assigned active network or if there is difference in the available network set
+         */
+        private fun processActiveNetwork(){
+            val activeNetwork: Network? = connectionMonitor.connectivityManager.activeNetwork
+            val isNewNetwork = connectionMonitor.currentActiveNetwork?.networkHandle != activeNetwork?.networkHandle
+            connectionMonitor.currentActiveNetwork = activeNetwork
+            if(!isNewNetwork){
+                return
+            }
+            if (activeNetwork != null) {
+                if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - process message ${connectionMonitor.connectivityManager.getNetworkInfo(activeNetwork)?.typeName.toString()}")
+                connectionMonitor.networkListener.onUpdateActiveNetwork(activeNetwork)
```

Comment:
ref comments on the previous diff, there's no need for a separate `onUpdateActiveNetwork`?

Reply:
Removed onUpdateActiveNetwork(). Now processActiveNetwork() will call onNetworkConnected()